src/coins.cpp

```diff
@@ -1195,18 +1228,49 @@ bool CCoinsViewCache::IsCertApplicableToState(const CScCertificate& cert, libzen
     const CFieldElement& scCumTreeHash_end   = curr_end_epoch_block_index->scCumTreeHash;
 
     // TODO Remove prev_end_epoch_block_hash after changing of verification circuit.
-    uint256 prev_end_epoch_block_hash = prev_end_epoch_block_index->GetBlockHash();
+    const uint256& prev_end_epoch_block_hash = prev_end_epoch_block_index->GetBlockHash();
 
     // Verify certificate proof
     CFieldElement constant{};
     if (sidechain.creationData.constant.is_initialized())
         constant = sidechain.creationData.constant.get();
+
     if (!scVerifier.verifyCScCertificate(constant, sidechain.creationData.wCertVk, prev_end_epoch_block_hash, cert))
     {
+        ret_code = RejectionCode::REJECT_PROOF_VER_FAILED;
         return error("%s():%d - ERROR: certificate[%s] cannot be accepted for sidechain [%s]: proof verification failed\n",
-                __func__, __LINE__, certHash.ToString(), cert.GetScId().ToString());
+            __func__, __LINE__, certHash.ToString(), cert.GetScId().ToString());
+    }
+
+    ret_code = RejectionCode::VALIDATION_OK;
+    return true;
+}
+
+bool CCoinsViewCache::CheckEndEpochCumScTxCommTreeRoot(
+    const CSidechain& sidechain, int epochNumber, const CFieldElement& endEpochCumScTxCommTreeRoot,
+	RejectionCode& ret_code) const
+{
+    LOCK(cs_main);
```

Comment:
I believe this could be an `AssertLockHeld(cs_main);` rather than lock seems all current callers already hold cs_main

Reply:
I would prefer leaving as it is because locks are re-entrant and also because AssertLockHeld expands to 'not null macro' only when DEBUG is activated